Standalone Soda Core guide

TJaniF opened the pull request:
Base PR for a potential standalone soda guide (to link to from the data quality overview guide) to make it easier to collaborate on.

denimalpaca commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,133 @@
+
+The bash script below:
+
+- creates a virtual environment
+- provides it with the `checks.yml` and `configuration.yml` files
+- enters the virtual environment
+- installs the `soda-core-snowflake` package in the virtual environment
+- runs the `soda scan` command on `MY_DATASOURCE` using the configuration files.
```
I think these should be updated to not have to install the package (or even enter the virtual environment). The bash script on a worker machine in Airflow should be lighter weight.

denimalpaca commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,133 @@
+      valid values: ['val1', 'val2', 'val3', 'val4']
+```
+
+After creation of the YAML files write a bash script that will be executed by the BashOperator in your Airflow environment. It is considered best practice to use Soda Core in a virtual environment.
````
Do we have a reference on where it says to use Soda Core in a virtual env? IIRC it was to install it in one, but given that we're just spinning up and down a worker node, it shouldn't matter because the env will be clean every time. And the lib should be installed in the Airflow instance's python path anyway.

TJaniF replied:
Yes, you are right, their documentation shows venv for local installation. I had issues at first running soda scan directly even though I had the lib installed at the OS level (the error I got was that the command `soda scan` was not available), but since it is working now I'll change it to not need a script at all. One less step 🙂

denimalpaca commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,133 @@
+- installs the `soda-core-snowflake` package in the virtual environment
+- runs the `soda scan` command on `MY_DATASOURCE` using the configuration files.
+
+> **Note** It is important to make this script executable by running `chmod +x filepath/soda_script.sh` before providing it to your Airflow environment. For more information on using bash scripts with Airflow see the ['Running scripts using the BashOperator' guide](https://www.astronomer.io/guides/scripts-bash-operator/).
```
`filepath` --> `<filepath>`

TJaniF replied:
Deleted this Note but was able to use that comment later :)

denimalpaca commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,133 @@
+> **Note** It is important to make this script executable by running `chmod +x filepath/soda_script.sh` before providing it to your Airflow environment. For more information on using bash scripts with Airflow see the ['Running scripts using the BashOperator' guide](https://www.astronomer.io/guides/scripts-bash-operator/).
+
+```bash
+#!/bin/bash
+python -m venv .venv
+cp /usr/local/airflow/include/soda/checks.yml .venv/include
+cp /usr/local/airflow/include/soda/configuration.yml .venv/include
+source .venv/bin/activate
+pip install --upgrade pip
+pip install soda-core-snowflake
+cd .venv/include
+soda scan -d MY_DATASOURCE -c configuration.yml checks.yml
````
With the `-c` flag in the soda scan command, you can specify the full file path. I think it would be better practice to specify a `SODA_PATH` env var in the Dockerfile and use that var in the command here. The `checks.yml` file should also be able to take a full file path.

Additionally, the `pip install` should happen in the `requirements.txt` file, or otherwise outside the script.

denimalpaca commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,133 @@
+        task_id="soda_test",
+        # please not the space at the end of the command!
+        bash_command="/usr/local/airflow/include/soda/soda_script.sh "
+    )
```
Make sure to add the task below the operator so it actually gets added to the DAG graph!

kentdanas replied:
I'm confused by this...unless you're using taskflow, instantiating an operator is enough to register it in the DAG. So you shouldn't need the `soda_test` line below, unless I'm missing something here...

TJaniF replied:
no, not missing anything. Benji and I were confused there 😅

denimalpaca reviewed (changes requested):
Looks pretty good overall, left comments about changes. Great work!

TJaniF commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,133 @@
+Reasons to use Soda Core are:
+
+- Very flexible in how you can define your checks using predefined metrics or full SQL statements.
+- [Offers check configurations](https://docs.soda.io/soda-cl/optional-config.html) that can add a warn state for a check. For example having values > 10 in a column might warrant a warning, while a value over 100 in the same column leads the check to fail.
```
True. I changed it to:

```markdown
- Lets you define checks as YAML configuration with many predefined metrics available.
- If there is no metric available for your use case you can provide any SQL query within the YAML file and check against its returned value.
- Easy to set up a large amount a data quality checks.
```

TJaniF replied:
Good point, thanks! :)

TJaniF commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,133 @@
+checks for example_table:
+  # check that MY_DATE_COL has dates between 2017-01-01 and 2022-01-01 using regex
+  - invalid_count(MY_DATE_COL) = 0:
+      valid regex: (((20)?(1[7-9]|2[0-1]))[-](0?[0-9]|1[012])[-](0?[0-9]|[12][0-9]|3[01])|2022[-]01[-]01)
```
thank you! 😅  I will be changing that check because it would only work if the date column is defined as TEXT which I hope for most users it would not be. I'll show of regex support with a less wieldy example.

denimalpaca replied:
An easier regex example might be looking for a certain website address or even just that a US postal code is formatted correctly.

(The regex is also still wrong in terms of date-checking because it does not conditionally check that a month is given an appropriate day, so it would assume the date `2021-02-30` is correct!)

TJaniF replied:
Yeah, also they do offer a whole list of shorthands including one for valid dates and postal codes: https://docs.soda.io/soda-cl/validity-metrics.html#list-of-valid-formats

The only reason this was a regex was because I ran a date bounds check on a date column in TEXT format because I   like to make my life harder apparently. 

certain website address is a good idea, I'll use that :)

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+- Easily set up a large amount of data quality checks.
+- Integrate data quality checks with commonly used data engineering tools such as Airflow, Apache Spark, PostgreSQL, Snowflake [and more](https://www.soda.io/integrations).
+
+> **Note**: For a general overview on how to approach data quality and of different tools available to run data quality checks using Airflow see the 'Data Quality and Airflow' guide.
```
```suggestion
> **Note**: For a general overview on how to approach data quality, and of different tools available to run data quality checks using Airflow, see the 'Data Quality and Airflow' guide.
```

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+- Provide any SQL query within the YAML file and check against its returned value, if no predefined metrics fit your use case.
+- Easily set up a large amount of data quality checks.
+- Integrate data quality checks with commonly used data engineering tools such as Airflow, Apache Spark, PostgreSQL, Snowflake [and more](https://www.soda.io/integrations).
+
```
Let's add an "In this guide, we'll cover how to use Soda Core with Airflow" type sentence here at the end of the overview.

kentdanas replied:
Yeah I think in general the more we can be consistent with the new templates the better, but I'll leave this one up to you however you prefer it.

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+---
+title: "Soda Core and Airflow"
+description: "Use Soda Core for data quality checks with Airflow"
```
```suggestion
description: "Using Soda Core to implement data quality checks in Airflow DAGs"
```

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+title: "Soda Core and Airflow"
+description: "Use Soda Core for data quality checks with Airflow"
+date: 2022-08-02T00:00:00.000Z
+slug: "data-quality"
```
The slug should be the same as the name of the file - so "soda_guide" in this case. Maybe we should make the name soda-data-quality or airflow-soda or something like that, to be a bit more consistent with the other ones (not that they all follow a convention, but there are a few patterns we tend to use...)

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+- How to design a data quality approach
+- What Airflow is and when to use it
+- Airflow Operators
```
```suggestion
- Airflow operators
```

kentdanas replied:
Once we have the update published, let's actually change this to the Airflow BashOperator specifically since we don't talk about any others, and then link to that guide.

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+To get the most out of this guide, users should have knowledge of:
+
+- How to design a data quality approach
+- What Airflow is and when to use it
```
I think we can leave this out of assumed knowledge sections for everything other than the "101" type guides. I tend to think if it would be in every single one then it isn't super meaningful

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+The following resources are recommended:
+
+- Data Quality and Airflow
+- [Introduction to Apache Airflow](Introduction to Apache Airflow)
```
```suggestion
```
Same comment as above, let's remove this one.

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+## Features of Soda Core
+
+Soda Core uses the SodaCL to run data quality checks defined in a YAML file. In this section we will highlight different types of checks. For a full overview please refer to the [Soda CL documentation](https://docs.soda.io/soda-cl/soda-cl-overview.html).
```
```suggestion
Soda Core uses the SodaCL to run data quality checks defined in a YAML file. In this section, we will highlight different types of checks you can implement with Soda Core. For a full overview please refer to the [Soda CL documentation](https://docs.soda.io/soda-cl/soda-cl-overview.html).
```

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+- [Relational database on Wikipedia](https://en.wikipedia.org/wiki/Relational_database)
+- [The Official YAML Web Site](https://yaml.org/)
+
+## Features of Soda Core
```
I think it would make sense somewhere in this section to add something about _why_ you would want to use Soda with Airflow

kentdanas commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,258 @@
+  - duplicate_count(MY_ID_COL) = 0
+```
+
+It is possible to add optional check configurations like naming a check or configure warning vs failing conditions:
````
```suggestion
It is possible to add optional check configurations like naming a check or configure warning versus failing conditions:
```

kentdanas commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,258 @@
+      name: Wrong number of rows!
+```
+
+How a metric is defined can be further specified using a:
````
```suggestion
How a metric is defined can be specified using any one of the following methods:
```

kentdanas commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,258 @@
+        WHERE MY_CATEGORY = 'category_1'
+```
+
+Two more unique features are freshness checks, schema checks and reference checks:
````
```suggestion
Two more unique features of Soda Core are freshness checks, schema checks and reference checks:
```
Shouldn't this say "three" more unique features? Or are "schema checks and reference checks" meant to be a single feature?

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+Soda Core uses the SodaCL to run data quality checks defined in a YAML file. In this section we will highlight different types of checks. For a full overview please refer to the [Soda CL documentation](https://docs.soda.io/soda-cl/soda-cl-overview.html).
+
+Standard check metrics offer the ability to run checks on individual columns:
```
Can we be a bit more explicit about what "metrics" are in this context?

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+## Requirements
+
+To use Soda Core, the Soda Core package for the database backend needs to be installed and configured. The [Soda documentation provides a list of supported databases](https://docs.soda.io/soda-core/configuration.html) and how to configure them.
```
```suggestion
To use Soda Core, the Soda Core package for the database backend needs to be installed and configured. The [Soda documentation provides a list of supported databases](https://docs.soda.io/soda-core/configuration.html) and how to configure them.
```

Is this a general requirement, or specific to using Soda Core with Airflow?

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+To use Soda Core, the Soda Core package for the database backend needs to be installed and configured. The [Soda documentation provides a list of supported databases](https://docs.soda.io/soda-core/configuration.html) and how to configure them.
+
+For the DAG code no additional providers or packages need to be installed since the `BashOperator` is part of the Airflow core.
```
Here we say no additional packages need to be installed, but in the section below one of the steps is to install the `soda-core-snowflake` package. Maybe we don't need a separate reqs section and we can just add a note somewhere that using Soda Core requires setting something up against your database?

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+- Create a configuration file that connects to the Snowflake database: `configuration.yml`.
+- Create a checks file containing the data quality checks: `checks.yml`.
+- Install the `soda-core-snowflake` package in your Airflow environment.
+- Create the DAG code running the `soca scan` command using the `BashOperator`.
```
```suggestion
- Run the checks from a DAG by running the `soca scan` command using the `BashOperator`.
```

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+For the DAG code no additional providers or packages need to be installed since the `BashOperator` is part of the Airflow core.
+
+## Example: Run Soda Core checks on a Snowflake database
```
```suggestion
## Example: Run Soda Core checks from Airflow
```

I think we should generally update the language in this section to be a bit more focused on Airflow.

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+### Step 1: Create the configuration file
+
+The configuration file needed to connect to Snowflake can be created from the template in the Soda documentation as shown below.
```
```suggestion
First we need to create a configuration file to connect to Snowflake. The easiest way to create the file is to use the template in the Soda documentation as shown below.
```

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+### Step 2: Create the checks file
+
+You can define your data quality checks using the [many checks available for Soda CL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a predefined metric or check that works for your use case you can create a user-defined check using SQL as shown below.
```
```suggestion
You can define your data quality checks using the [many checks available for Soda CL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a predefined metric or check that works for your use case, you can create a user-defined check using SQL as shown below.
```

kentdanas commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,258 @@
+      valid values: ['val1', 'val2', 'val3', 'val4']
+```
+
+> **Note**: You need to make both YAML files available to your Airflow environment. Users of the Astro CLI can simply put them into the `/include` folder.
````
I was looking for this comment a bit earlier, so let's clearly state in both sections where to put the YAML files.

TJaniF replied:
put it into requirements too 🙂

kentdanas commented on guides/soda_guide.md:
```diff
@@ -0,0 +1,258 @@
+
+### Step 4: Run soda scan using the BashOperator
+
+The DAG code itself stays very simple with one task using the `BashOperator` to execute the `soda scan` command. Of course it is possible to incorporate data quality checks using Soda Core in different locations within your data pipeline.  
```
```suggestion
In a DAG, Soda Core checks are executed by using the `BashOperator` to run the `soda scan` command. The DAG below shows how to reference the configuration and checks YAML files in the command.
```

kentdanas commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,258 @@
+  File "/usr/local/lib/python3.9/site-packages/airflow/operators/bash.py", line 194, in execute
+    raise AirflowException(
+airflow.exceptions.AirflowException: Bash command failed. The command returned a non-zero exit code 2.
+```
````
```suggestion
```

Of course it is possible to incorporate data quality checks using Soda Core in different locations within your DAG, and to use the results of those checks to control the flow of data within your pipeline.  
```

kentdanas commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,258 @@
+    soda_test
+```
+
+The logs from the Soda Core checks can be found in the Airflow task logs. They will list all checks that ran with their result.
````
```suggestion
The logs from the Soda Core checks can be found in the Airflow task logs. The logs will list all checks that ran along with their result.
```

kentdanas commented on guides/soda_guide.md:
````diff
@@ -0,0 +1,258 @@
+[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO - All is good. No failures. No warnings. No errors.
+```
+
+In case of failure of a check, the logs will show which check failed and what the `check_value` was that caused the failure.
````
```suggestion
In the case of a check failure, the logs will show which check failed and the `check_value` that caused the failure.
```

TJaniF reviewed (commented):
Addressed the comments 🙂 . I also changed the first check in the example to not be a regex since the option to use regex is now shown in the features section and since regex only works on TEXT columns I don't think many users would need to run that on dates.

kentdanas commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,262 @@
+The following resources are recommended:
+
+- Data Quality and Airflow
+- [Operators 101](Operators 101)
```
Update to the link to the guide here

TJaniF replied:
I could have sworn I did correct that... weird. Sorry! 😅

kentdanas commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,262 @@
+
+Data can be filtered according to validity criteria using the following methods:
+
+- List of valid values.
```
I think we don't need the periods at the end of these bullet points since they aren't full sentences

kentdanas commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,262 @@
+
+To use Soda Core, the Soda Core package for the database backend needs to be installed and configured. The [Soda documentation provides a list of supported databases](https://docs.soda.io/soda-core/configuration.html) and how to configure them.
+
+When using Soda Core with Airflow the Soda Core package needs to be installed in your Airflow environment. Additionally, both YAML files, `configuration.yml`, which contains the connection information for the database backend and `checks.yml`, which contains the data quality checks will need to be made available to the Airflow environment.
```
```suggestion
When using Soda Core with Airflow, the Soda Core package needs to be installed in your Airflow environment. Additionally, two YAML files need to be made available to the Airflow environment: `configuration.yml`, which contains the connection information for the database backend, and `checks.yml`, which contains the data quality checks.
```

kentdanas commented on guides/soda-data-quality.md:
````diff
@@ -0,0 +1,262 @@
+      QUOTED_IDENTIFIERS_IGNORE_CASE: false
+  schema: MY_SCHEMA
+```
+
````
Let's add a sentence here that says something like "Paste the text above into a file called `configuration.yml` and place it in your `include/` directory"

kentdanas commented on guides/soda-data-quality.md:
````diff
@@ -0,0 +1,262 @@
+      valid values: ['val1', 'val2', 'val3', 'val4']
+```
+
+> **Note**: You need to make both YAML files available to your Airflow environment. Users of the Astro CLI can simply put them into the `/include` folder.
````
Same comment here as above - instead of having a note about this, let's add an explicit sentence in both sections saying to put the text in a file called xyz and put it somewhere where it will be available to Airflow e.g. include/

kentdanas reviewed (approved):
Looks great @TJaniF! A couple of minor comments, but otherwise we're good to move this forward to copy edit.

TJaniF reviewed (commented):
Thank you! :) @jwitz this guide was kinda unplanned and happened because I wrote too much Soda content for the general data quality guide 😅 . Has the lowest priority of the reviews. 🙂

jwitz commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,257 @@
+To use Soda Core, the Soda Core package for the database backend needs to be installed and configured. The [Soda documentation provides a list of supported databases](https://docs.soda.io/soda-core/configuration.html) and how to configure them.
+
+When using Soda Core with Airflow, the Soda Core package needs to be installed in your Airflow environment. Additionally, two YAML files need to be made available to the Airflow environment: `configuration.yml`, which contains the connection information for the database backend, and `checks.yml`, which contains the data quality checks.
+Astro CLI users can simply put `soda-core-<database>` in `requirements.txt` and provide the YAML files in `/include`.
```
```suggestion
If you use the Astro CLI, you can add `soda-core-<database>` to `requirements.txt` and your YAML files to `/include` in your Astro project.
```

jwitz commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,257 @@
+
+To use Soda Core, the Soda Core package for the database backend needs to be installed and configured. The [Soda documentation provides a list of supported databases](https://docs.soda.io/soda-core/configuration.html) and how to configure them.
+
+When using Soda Core with Airflow, the Soda Core package needs to be installed in your Airflow environment. Additionally, two YAML files need to be made available to the Airflow environment: `configuration.yml`, which contains the connection information for the database backend, and `checks.yml`, which contains the data quality checks.
```
```suggestion
When using Soda Core with Airflow, the Soda Core package needs to be installed in your Airflow environment. Additionally, two YAML files need to be made available to the Airflow environment: `configuration.yml`, which contains the connection information for the database backend, and `checks.yml`, which contains the data quality checks.

```

jwitz commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,257 @@
+
+## Example: Run Soda Core checks from Airflow
+
+This example shows one possible way to set up data quality checks on a Snowflake database using Soda Core from within Airflow.
```
```suggestion
This example shows one possible way to run data quality checks on a Snowflake database using Soda Core from within Airflow.
```

jwitz commented on guides/soda-data-quality.md:
````diff
@@ -0,0 +1,257 @@
+  schema: MY_SCHEMA
+```
+
+Save the YAML instructions above in a filed called `configuration.yml` and make it available to your Airflow environment. Astro CLI users can place the file into the `/include` directory.
````
```suggestion
Save the YAML instructions above in a filed called `configuration.yml` and make it available to your Airflow environment. If you use the Astro CLI, you can do this by placing the file into the `/include` directory of your Astro project.
```

jwitz commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,257 @@
+
+### Step 2: Create the checks file
+
+You can define your data quality checks using the [many checks available for Soda CL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a predefined metric or check that works for your use case, you can create a user-defined check using SQL as shown below.
```
```suggestion
You can define your data quality checks using the [many preset checks available for Soda CL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a preset metric or check that works for your use case, you can create a custom check using SQL as shown below.
```
Still confused about metric vs. check here

TJaniF replied:
Yeah, that is on me, sorry! I misread something and and thought things like missing_count or duplicate_count were called metrics and other checks were called checks. But as I said above metrics are the property of the data that is being checked. I'll change it in the text. 🙂

jwitz commented on guides/soda-data-quality.md:
````diff
@@ -0,0 +1,257 @@
+      valid values: ['val1', 'val2', 'val3', 'val4']
+```
+
+Save the YAML instructions above in a filed called `checks.yml` and make it available to your Airflow environment. Astro CLI users can place the file into the `/include` directory.
````
```suggestion
Save the YAML instructions above in a filed called `checks.yml` and make it available to your Airflow environment. If you use the Astro CLI, you can place this file in your `/include` directory.
```

jwitz commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,257 @@
+
+### Step 3: Install the Soda Core package
+
+The `soda-core-snowflake` package needs to be installed in your Airflow environment. Astro CLI users can add the following line to the `requirements.txt` file:
```
```suggestion
The `soda-core-snowflake` package needs to be installed in your Airflow environment. If you use the Astro CLI, you can do this by adding the following line to the `requirements.txt` file:
```

jwitz commented on guides/soda-data-quality.md:
````diff
@@ -0,0 +1,257 @@
+    )
+```
+
+The logs from the Soda Core checks can be found in the Airflow task logs. The logs will list all checks that ran along with their result.
````
```suggestion
Because Soda Core runs through the BashOperator, you can run your checks in any part of your DAG and trigger tasks with your results like you would with other Airflow operators.

The logs from the Soda Core checks can be found in the Airflow task logs. The logs list all checks that ran and their results.
```

jwitz commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,257 @@
+
+The logs from the Soda Core checks can be found in the Airflow task logs. The logs will list all checks that ran along with their result.
+
+Below an example of the logs in the case of 3 checks passing:
```
```suggestion
The following is an example of how your logs might look if all 3 checks pass:
```

jwitz commented on guides/soda-data-quality.md:
````diff
@@ -0,0 +1,257 @@
+[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO - All is good. No failures. No warnings. No errors.
+```
+
+In the case of a check failure, the logs will show which check failed and the `check_value` that caused the failure.
````
```suggestion
In the case of a check failure, the logs show which check failed and the `check_value` that caused the failure:
```

jwitz commented on guides/soda-data-quality.md:
````diff
@@ -0,0 +1,257 @@
+airflow.exceptions.AirflowException: Bash command failed. The command returned a non-zero exit code 2.
+```
+
+Of course it is possible to incorporate data quality checks using Soda Core in different locations within your DAG, and to use the results of those checks to control the flow of data within your pipeline.
````
```suggestion
```
If you want to keep this sentence, I think it's better to add it before the DAG code above. It feels a bit like a cliffhanger here.

TJaniF replied:
🧗

jwitz reviewed (approved):
LGTM! Just a few minor suggestions

janet-can commented:
Hi @TJaniF  - I'm the Tech Writer at Soda and I'm really excited to see your work putting together this material for Astronomer users. I don't know if it would be useful to you, but I would be very happy to collaborate on any of the details, or to address anything in our own documentation that warrants further clarification or expansion. Please do reach out if I can help! janet@soda.io

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,257 @@
+
+Soda Core lets you:
+
+- Define checks as YAML configuration with many predefined metrics available.
```
No, you are correct, metric is the property of the dataset that the check is run on. Thanks for catching that 😅

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,257 @@
+
+- How to design a data quality approach. See Data Quality and Airflow.
+- Airflow operators. See [Operators 101](https://www.astronomer.io/guides/what-is-an-operator/).
+- Relational Databases. See [Relational database on Wikipedia](https://en.wikipedia.org/wiki/Relational_database).
```
True! I wasn't sure what our policy is with linking to resources on general concepts from companies, that is why I went with wikipedia, but yours is better.

TJaniF reviewed (commented):
Thank you for the comments @jwitz! And happy to hear your thoughts @janet-can 🙂 !

janet-can commented:
I had a read-through of the guide and it's terrific! I don't really have a whole lot to add, but I made a few notes for your consideration.

- Potentially add [How Soda Core works](https://docs.soda.io/soda-core/how-core-works.html) to the list of Assumed Knowledge items, in case it's easy to have a quick ref for what Soda is and does.
- In the first paragraph under Features of Soda Core, "Soda Core uses SodaCL to run data quality checks..."  It might be helpful to expand the acronym on first mention, "Soda Checks Language (SodaCL)".
- In the fifth paragraph under Features of Soda Core, "Data can be filtered according to validity criteria..." Hmmm, this isn't really a filter. The invalid check counts all rows with contents that do not match the valid values listed and returns that count in the check result, for example; and the missing check counts everything that doesn't match an IP address format and returns that count in the check result.  Maybe just as simple as replacing the word "filtered" with "checked". (This is a nit-picky detail, I know!)
- In the Step 2 example, a very minor detail: in a valid values list, only numeric values need to be wrapped in single-quotes; strings can just be a plain comma-separated list. 
```# checks that all entries in MY_COL_3 are part of a set or possible values
  - invalid_count(MY_COL_3) = 0:
      valid values: ['val1', 'val2', 'val3', 'val4']
```  

Marvellous work, if I may say so!  When you publish this, I'd love to announce it in our community Slack channel, if that's OK with you. More eyes on it will be a good thing :)

TJaniF commented:
Thank you for the notes and feedback @janet-can, highly appreciate your help! 🙂 

Those are all great suggestions. I didn't realize that you needed quotes only for numeric values (I see it now in your docs but somehow I must have missed that line previously 😅 ), to point it out I added another valid values check for numeric values right underneath the existing check in the Features section (line 76-83). 

And yes of course you can announce the published guide in your Slack! Always happy to reach more people! 🙂 

@kentdanas I think this guide is ready to be published after the data quality overview one, if you agree with the latest changes. 🙂 (I'd dismiss Benjis review since I addressed his comments and I don't think we need to wait for the end of his vacation)

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+
+To get the most out of this guide, you should have knowledge of:
+
+- How to design a data quality approach. See Data Quality and Airflow.
```
```suggestion
- How to design a data quality approach. See [Data Quality and Airflow](https://www.astronomer.io/guides/data-quality).
```

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+
+In this guide, we will cover key features of Soda Core and how to use it with Airflow.
+
+> **Note**: For a general overview on how to approach data quality, and of different tools available to run data quality checks using Airflow, see the 'Data Quality and Airflow' guide.
```
```suggestion
> **Note**: For a general overview on how to approach data quality, and of different tools available to run data quality checks using Airflow, see the '[Data quality and Airflow](https://www.astronomer.io/guides/data-quality)' guide.
```

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+---
+title: "Soda Core and Airflow"
+description: "Using Soda Core to implement data quality checks in Airflow DAGs"
```
```suggestion
description: "Using Soda Core to implement data quality checks in Airflow DAGs."
```

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+---
+title: "Soda Core and Airflow"
+description: "Using Soda Core to implement data quality checks in Airflow DAGs"
+date: 2022-08-02T00:00:00.000Z
```
```suggestion
date: 2022-08-11T00:00:00.000Z
```

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+date: 2022-08-02T00:00:00.000Z
+slug: "soda-data-quality"
+heroImagePath: null
+tags: ["Soda", "Data Quality"]
```
```suggestion
tags: ["Soda", "Data quality"]
```

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+  - schema:
+      fail:
+        when required column missing: [MY_KEY]
+  # all names listed in MY_TABLE_1's MY_NAMES column have to also exists
```
```suggestion
  # all names listed in MY_TABLE_1's MY_NAMES column have to also exist
```

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+- Install the `soda-core-snowflake` package in your Airflow environment.
+- Run the checks from a DAG by running the `soda scan` command using the `BashOperator`.
+
+
```
```suggestion
```

TJaniF commented on guides/soda-data-quality.md:
````diff
@@ -0,0 +1,267 @@
+  schema: MY_SCHEMA
+```
+
+Save the YAML instructions above in a filed called `configuration.yml` and make it available to your Airflow environment. If you use the Astro CLI, you can do this by placing the file into the `/include` directory of your Astro project.
````
```suggestion
Save the YAML instructions above in a file called `configuration.yml` and make it available to your Airflow environment. If you use the Astro CLI, you can do this by placing the file into the `/include` directory of your Astro project.
```

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+
+### Step 2: Create the checks file
+
+You can define your data quality checks using the [many preset checks available for Soda CL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a preset check that works for your use case, you can create a custom one using SQL as shown below.
```
```suggestion
You can define your data quality checks using the [many preset checks available for SodaCL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a preset check that works for your use case, you can create a custom one using SQL as shown below.
```

TJaniF commented on guides/soda-data-quality.md:
```diff
@@ -0,0 +1,267 @@
+
+The logs from the Soda Core checks can be found in the Airflow task logs. The logs list all checks that ran and their results.
+
+The following is an example of how your logs might look if all 3 checks pass:
```
```suggestion
This is an example of what your logs might look like if all 3 checks pass:
```
